### ISSUE TITLE
Numerical reproducibility in saving/loading an xp.Particle after a tracking step in a xt.Cavity

### ISSUE THREAD

**Issue by carlidel** [xsuite/xtrack#85]
@giadarol and I recently inspected the following bug.

Consider the following script:
```python
import numpy as np

import xobjects as xo
import xtrack as xt
import xpart as xp

def areIdentical(part1, part2):
    flag = True
    for tt, nn in xp.Particles._structure['scalar_vars']:
        if not getattr(part1, nn) == getattr(part2, nn):
            print(f"not identical on {nn}")
            flag = False
    for tt, nn in xp.Particles._structure['per_particle_vars']:
        if not np.all(getattr(part1, nn) == getattr(part2, nn)):
            print(f"not identical on {nn}")
            print(f"part1: {getattr(part1, nn)}")
            print(f"part2: {getattr(part2, nn)}")
            flag = False
    return flag

context = xo.ContextCpu()

samples = 2
extent = (0.1e-4, 1e-4)
x = np.linspace(*extent, samples)
xx, yy = np.meshgrid(x, x)
x_f = xx.flatten()
y_f = yy.flatten()
px_f = np.zeros_like(x_f)
py_f = np.zeros_like(x_f)


particles = xp.Particles(
    _context=context, p0c=6500e9,
    x=x_f, px=px_f, y=y_f, py=py_f, zeta=1e-12, delta=1e-9,
)

print("Check self-consistency:")
dct = particles.copy().to_dict()
pfromdict = xp.Particles.from_dict(dct)
print(areIdentical(particles, pfromdict))

cav = xt.Cavity(voltage=1e6, frequency=400e6)
cav.track(particles)

print("Check self-consistency after one tracking step in Cavity:")
dct = particles.copy().to_dict()
pfromdict = xp.Particles.from_dict(dct)
print(areIdentical(particles, pfromdict))
```

The expected behavior is to get `True` on both instances. As this would allow the consistent creation of checkpoint savefiles for very long tracking jobs.

Instead, we get the following output:

```
Check self-consistency:
True
Check self-consistency after one tracking step in Cavity:
not identical on psigma
part1: [9.99999882e-10 9.99999882e-10 9.99999882e-10 9.99999882e-10]
part2: [9.99999871e-10 9.99999871e-10 9.99999871e-10 9.99999871e-10]
False
```

i.e. some internal methods in the initialization of `xp.Particles.from_dict` seem to do some internal normalization of the longitudinal dynamics that could lead to a very little numerical fluctuation.

An initial investigation lead by @giadarol highlighted a possible culprit in this part of code from `xpart/xpart/particles/_pyparticles.py` (lines 113-138): 
```python

    def __init__delta(self, delta, ptau, psigma):
        not_none = count_not_none(delta, ptau, psigma)
        if not_none == 0:
            self.delta = 0.0
        elif not_none >= 1:
            if delta is not None:
                self.delta = delta
            elif ptau is not None:
                self.ptau = ptau
            elif psigma is not None:
                self.psigma = psigma

            if not_none>1:
                ddd = {'delta': delta, 'ptau': ptau, 'psigma': psigma}
                for nn, vv in ddd.items():
                    if vv is None:
                        continue

                    if not np.allclose(vv, getattr(self, nn), atol=1e-13):
                        raise ValueError(
                            f"""
                        Particles defined with inconsistent energy deviations:
                        delta  = {delta},
                        ptau     = {ptau},
                        psigma = {psigma}"""
                        )
```
but the main issue is yet to be found and corrected.

**Comment by giadarol** [comment]
This branch should fix the issue:
https://github.com/xsuite/xpart/tree/fix/forceenevars

**Comment by giadarol** [comment]
Solved in Xsuite v0.2.1
